pkg/handler/v2/external_cluster/kubeone.go

```diff
@@ -74,55 +79,165 @@ func importKubeOneCluster(ctx context.Context, name string, userInfoGetter func(
+	}
 	manifestSecret := &corev1.Secret{}
 	if err := masterClient.Get(ctx, types.NamespacedName{Namespace: manifest.Namespace, Name: manifest.Name}, manifestSecret); err != nil {
 		return nil, errors.NewBadRequest(fmt.Sprintf("can not retrieve kubeone manifest secret: %v", err))
 	}
-	oldManifestSecret := manifestSecret.DeepCopy()
+	currentManifest := manifestSecret.Data[resources.KubeOneManifest]
+	cluster := &kubeonev1beta2.KubeOneCluster{}
+	if err := yaml.UnmarshalStrict(currentManifest, cluster); err != nil {
+		return nil, fmt.Errorf("failed to decode manifest secret data: %w", err)
+	}
+	cluster.ContainerRuntime.Docker = nil
+	cluster.ContainerRuntime.Containerd = &kubeonev1beta2.ContainerRuntimeContainerd{}
 
+	patchManifest, err := yaml.Marshal(cluster)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode kubeone cluster manifest config as YAML: %w", err)
+	}
+
+	oldManifestSecret := manifestSecret.DeepCopy()
 	manifestSecret.Data = map[string][]byte{
-		resources.KubeOneManifest: yamlManifest,
+		resources.KubeOneManifest: patchManifest,
 	}
 	if err := masterClient.Patch(ctx, manifestSecret, ctrlruntimeclient.MergeFrom(oldManifestSecret)); err != nil {
-		return nil, fmt.Errorf("failed to update kubeone manifest secret for upgrade version %s/%s: %w", manifest.Name, manifest.Namespace, err)
+		return nil, fmt.Errorf("failed to update kubeone manifest secret for container-runtime containerd %s/%s: %w", manifest.Name, manifest.Namespace, err)
 	}
 
 	oldexternalCluster := externalCluster.DeepCopy()
 	// update kubeone externalcluster status.
 	externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.Status = kubermaticv1.StatusReconciling
-	if upgradeMD == "true" {
-		externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.StatusMessage = "upgrading machine MD"
-	}
-	// update api externalcluster status
-	newCluster.Status.State = apiv2.RECONCILING
+	// update api externalcluster status.
+	newCluster.Status = apiv2.ExternalClusterStatus{State: apiv2.RECONCILING}
+
 	if err := masterClient.Patch(ctx, externalCluster, ctrlruntimeclient.MergeFrom(oldexternalCluster)); err != nil {
 		return nil, fmt.Errorf("failed to update kubeone cluster status %s: %w", externalCluster.Name, err)
 	}
 
 	return newCluster, nil
 }
+
+func checkContainerRuntime(ctx context.Context,
+	externalCluster *kubermaticv1.ExternalCluster,
+	externalClusterProvider provider.ExternalClusterProvider,
+) (string, error) {
+	var containerRuntime string
+	nodes, err := externalClusterProvider.ListNodes(ctx, externalCluster)
+	if err != nil {
+		return "", fmt.Errorf("failed to list nodes: %w", err)
+	}
+	for _, node := range nodes.Items {
+		if _, ok := node.Labels[NodeControlPlaneLabel]; ok {
+			containerRuntimeVersion := node.Status.NodeInfo.ContainerRuntimeVersion
+			strSlice := strings.Split(containerRuntimeVersion, ":")
+			for _, v := range strSlice {
+				containerRuntime = v
```

Comment:
why just return here

Reply:
maybe return strSlice[0]

Reply:
i just did this to avoid the case when the slice is empty, indexnotfound err.
but yes, we can return here.

Reply:
otherwise, you will return an empty string. Is it OK?

Reply:
we can return an err, in case of empty string.
wdyt?

Reply:
err could be something like no container runtime found...

Reply:
err or some default, I am not sure

Reply:
the error will be better